File: src/waypalm-site/src/main/java/waypalm/site/web/extensions/Extender.java
package waypalm.site.web.extensions;

import org.springframework.beans.BeanUtils;
import org.springframework.ui.ModelMap;
import org.springframework.web.context.request.WebRequest;
import waypalm.common.web.extensions.ModelExtender;
import waypalm.site.services.UserAccess;

import javax.inject.Inject;

public abstract class Extender<T> implements ModelExtender {
    @Inject
    protected UserAccess auth;

    @SuppressWarnings("unchecked")
    @Override
    public final void extend(String extensionKey, WebRequest request, ModelMap model) {
        Object extensionInstance = model.get(extensionKey);
        Object extensionApplied = createExtension(request, (T) extensionInstance, model);
        if (extensionApplied != null) {
            model.put(extensionKey, extensionApplied);
        } else {
            model.remove(extensionKey);
        }
    }

    public T createExtension(WebRequest request, T extension, ModelMap model) {
        return populate(request, extension, model);
    }

    protected abstract T populate(WebRequest request, T extension, ModelMap model);

    protected <T> T instantiateIfNull(T value, Class<T> valueClass) {
        if (value == null) {
            value = BeanUtils.instantiate(valueClass);
        }
        return value;
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/web/extensions/Extender.kt
package waypalm.site.web.extensions

import waypalm.common.web.extensions.ModelExtender
import org.springframework.web.context.request.WebRequest
import org.springframework.ui.ModelMap
import org.springframework.beans.BeanUtils

public abstract class Extender<T> ()
: ModelExtender
{
    public override fun extend(extensionName: String, request: WebRequest, model: ModelMap) {
        val extensionInstance = model.get(extensionName);
        val extensionApplied = createExtension(request, extensionInstance as T, model);
        if (extensionApplied != null) {
            model.put(extensionName, extensionApplied);
        } else {
            model.remove(extensionName);
        }
    }

    public fun  createExtension(request: WebRequest, extension: T?, model: ModelMap): T? {
        return populate(request, extension, model);
    }

    protected abstract fun populate(request: WebRequest, extension: T?, model: ModelMap): T?;

    protected fun instantiateIfNull<C>(value: C?, valueClass: Class<C>): C {
        return if (value != null) value else BeanUtils.instantiate(valueClass)!!
    }
}